File: references/eval/failures.py
#!/usr/bin/env python3
"""
Failure diagnosis module for model evaluation.
Classifies WHY a model result failed, and carries the remedy for the failure modes
whose remedy is a harness setting rather than a prompt change.
"""

import os
import re

from lib.validators_lib import has_item_details

from eval.validate import safe_content

FAIL_INFRA = "INFRA"
FAIL_TIMEOUT = "TIMEOUT"
FAIL_PARSE = "PARSE"
FAIL_FORMAT = "FORMAT"
FAIL_CONTENT = "CONTENT"
# A reasoning model that spent its whole budget thinking and returned nothing. Kept
# separate from FORMAT because the remedy is opposite: FORMAT failures want a clearer
# prompt, this wants a SMALLER token budget so the model is forced to stop.
FAIL_REASONING = "REASONING"
FAIL_NONE = None

# HTTP 5xx, or an explicit overload marker. Matched on the error string because
# that is all the transport hands back once a request has failed.
_SERVER_ERROR_RE = re.compile(
    r"HTTP\s+5\d\d|server_overloaded|inference capacity|service unavailable",
    re.IGNORECASE,
)


def _classify_failure(result: dict, task_cfg: dict, score: int, failure_reason: str) -> dict:
    """Classify WHY a result failed. Returns a diagnosis dict.

    Returns:
        {
            "category": one of FAIL_* constants,
            "reason": human-readable failure reason,
            "evidence": specific evidence for the diagnosis,
        }
    """
    from eval.run import DEFAULT_EVAL_TIMEOUT

    error = result.get("error") or ""
    content = safe_content(result)
    parsed = result.get("parsed")
    raw_len = len(content)

    if score >= 90:
        return {"category": FAIL_NONE, "reason": "", "evidence": ""}

    if "Model not found" in error:
        return {
            "category": FAIL_INFRA,
            "reason": error,
            "evidence": "Model not loaded or wrong identifier",
        }
    if "Connection" in error:
        return {
            "category": FAIL_INFRA,
            "reason": error,
            "evidence": "Server unreachable",
        }

    # An HTTP 5xx is the SERVER failing, never the model producing bad output.
    # These were landing in FORMAT, so qwen3.6-35b's 34 "HTTP 503: Server is at
    # inference capacity" responses were recorded as 34 formatting failures --
    # a server outage written down as a model's quality score, and invisible to
    # anything looking for infrastructure trouble.
    if _SERVER_ERROR_RE.search(error):
        return {
            "category": FAIL_INFRA,
            "reason": error,
            "evidence": "Server returned 5xx; the model never got to answer",
        }

    if "Timeout" in error or "timed out" in error.lower():
        return {
            "category": FAIL_TIMEOUT,
            "reason": error,
            "evidence": f"Model did not respond within {DEFAULT_EVAL_TIMEOUT}s",
        }

    # Checked BEFORE the parse_json branch. A reasoning model that never stopped
    # returns empty content, and on a JSON task that reads as "no JSON brackets" --
    # the same mislabel one level down, and the likelier one, since the weekend tasks
    # are the hard prompts that trigger it.
    #
    # The qwen3_5 family streams its chain of thought into `reasoning_content` and
    # leaves `content` empty until it stops. Calling that a FORMAT failure reads as a
    # model that cannot follow instructions, so the response is to rewrite the prompt
    # -- when the fix is to make it STOP reasoning. Raising the budget makes it
    # strictly worse: the reasoning expands to fill whatever it is given.
    reasoning = (result or {}).get("reasoning_content") or ""
    if not content and reasoning:
        finish = (result or {}).get("finish_reason") or "unknown"
        return {
            "category": FAIL_REASONING,
            "reason": "Reasoned past the token budget",
            "evidence": (
                f"{len(reasoning)} chars of reasoning_content, empty content, "
                f"finish_reason={finish}. Not a prompt-following failure: the model "
                f"never stopped thinking, so raise max_tokens for this task. "
                f"Reasoning scales with the TASK, not with the budget: nemotron "
                f"needs ~19k chars of reasoning on an 8-line prompt and ~77k on a "
                f"40-line one, and returns empty at every budget below what it "
                f"needs. Shrinking the budget does NOT force an answer."
            ),
        }

    if task_cfg["parse_json"]:
        has_json_chars = "{" in content or "[" in content
        has_prose_before_json = False
        if has_json_chars:
            first_bracket = min(
                (content.find(c) for c in "[{" if content.find(c) >= 0),
                default=raw_len,
            )
            has_prose_before_json = first_bracket > 200

        if not parsed and not has_json_chars:
            return {
                "category": FAIL_FORMAT,
                "reason": failure_reason or "No JSON in output",
                "evidence": f"Output was {raw_len} chars of prose with no JSON brackets",
            }

        if not parsed and has_json_chars:
            return {
                "category": FAIL_PARSE,
                "reason": failure_reason or "JSON extraction failed",
                "evidence": (
                    f"Output had JSON-like content at char {first_bracket} of "
                    f"{raw_len} but parser couldn't extract valid JSON"
                ),
            }

        if parsed and has_prose_before_json:
            evidence = (
                f"Model emitted {first_bracket} chars of reasoning before "
                f"first JSON bracket (total {raw_len} chars)"
            )
            if score < 50:
                return {
                    "category": FAIL_FORMAT,
                    "reason": failure_reason,
                    "evidence": evidence + "; reasoning may have consumed the context window",
                }
            return {
                "category": FAIL_CONTENT,
                "reason": failure_reason,
                "evidence": evidence,
            }

        return {
            "category": FAIL_CONTENT,
            "reason": failure_reason,
            "evidence": _describe_content_failure(parsed, failure_reason),
        }

    if not content:
        return {
            "category": FAIL_FORMAT,
            "reason": "Empty content",
            "evidence": "Model returned empty response",
        }

    reasoning_markers = ["Let me", "I'll", "Wait,", "Actually,", "Here's my", "Thinking"]
    has_reasoning = any(marker in content[:200] for marker in reasoning_markers)
    if has_reasoning and raw_len > 200:
        return {
            "category": FAIL_FORMAT,
            "reason": failure_reason,
            "evidence": (
                f"Model output {raw_len} chars starting with reasoning "
                f"instead of a direct answer"
            ),
        }

    return {
        "category": FAIL_CONTENT,
        "reason": failure_reason,
        "evidence": f"Output was {raw_len} chars but failed validation: {failure_reason}",
    }


def _describe_content_failure(parsed, failure_reason: str) -> str:
    """Generate human-readable evidence for content quality failures."""
    if isinstance(parsed, list):
        item_count = len(parsed)
        with_details = sum(
            1 for item in parsed if isinstance(item, dict) and has_item_details(item)
        )
        return f"Parsed {item_count} items, {with_details} with details. {failure_reason}"
    elif isinstance(parsed, dict):
        keys = list(parsed.keys())[:5]
        return f"Parsed dict with keys {keys}. {failure_reason}"
    return failure_reason


# Multiplier for a retry after a reasoning overrun. An overrun means the model wanted
# MORE room, so the retry gives it more.
#
# This used to retry at a flat 512, on the theory that a tight budget forces a model
# to stop thinking and answer. That generalised from one model and is false for
# others: nemotron-3.5-lightning returns empty at 512, 2048, 8192 and 16000 alike --
# guard on or off -- and answers only at 32000, having emitted 77,208 characters of
# reasoning first. Against that behaviour a 512-token retry does not force an answer,
# it guarantees a zero, and the zero then averages into the model's quality mean as
# though it had answered badly.
REASONING_RETRY_MULTIPLIER = float(os.environ.get("EVAL_REASONING_RETRY_MULT", "2.0"))
# Ceiling so an escalating retry cannot run away. 64000 is 2x the largest configured
# budget and inside every installed server model's context window (131k-262k).
REASONING_RETRY_MAX_TOKENS = int(os.environ.get("EVAL_REASONING_RETRY_TOKENS", "64000"))


def reasoning_retry_budget(base_budget: int) -> int:
    """Token budget for a retry after a reasoning overrun: MORE room, bounded.

    Lives here rather than in run.py because it is the remedy for FAIL_REASONING,
    and keeping the diagnosis and its remedy apart is how they came to disagree --
    the evidence string above told the reader to shrink the budget while the retry
    was shrinking it for them, and both were wrong.
    """
    return min(int(base_budget * REASONING_RETRY_MULTIPLIER), REASONING_RETRY_MAX_TOKENS)

File: rust/src/ztools/eval/failures.rs
//! Failure classification for the eval loop: WHY a result failed.
//!
//! Ported from `eval/failures.py::_classify_failure`. The category decides
//! what happens NEXT, so a mislabel is not cosmetic:
//!
//! - INFRA (server 5xx, unreachable, missing model) and TIMEOUT count toward
//!   abandoning a model whose SERVER is failing -- those zeros are not quality
//!   results. This split is why qwen3.6-35b's 34 "HTTP 503" responses are an
//!   outage and not 34 formatting failures.
//! - REASONING (`reasoning_content` present, content empty) triggers a retry
//!   with MORE room: the model never stopped thinking, and reasoning scales
//!   with the TASK, not the budget. Checked BEFORE any JSON/FORMAT branch --
//!   on a JSON task an endless thinker reads as "no JSON brackets", which used
//!   to be answered by prompt rewrites instead of budget.
//! - PARSE marks a parse failure in the signal store, distinct from a model
//!   that answered wrong.

use crate::units::whole_u32;
use serde_json::Value;

pub const FAIL_INFRA: &str = "INFRA";
pub const FAIL_TIMEOUT: &str = "TIMEOUT";
pub const FAIL_PARSE: &str = "PARSE";
pub const FAIL_FORMAT: &str = "FORMAT";
pub const FAIL_CONTENT: &str = "CONTENT";
pub const FAIL_REASONING: &str = "REASONING";
/// No failure. Python uses None; Rust uses an empty category string.
pub const FAIL_NONE: &str = "";

/// Retry budget after a REASONING overrun: MORE room, bounded. Lives next to
/// the classifier because keeping the diagnosis and its remedy apart is how
/// they came to disagree once already.
pub const REASONING_RETRY_MULTIPLIER: f64 = 2.0;
pub const REASONING_RETRY_MAX_TOKENS: u32 = 64_000;

/// The `finish_reason` the stream guard stamps when it cuts a run that has
/// reasoned past the point where the remaining budget could hold an answer.
///
/// Set in eval/transport.rs (and mirrored in Python lib/llm/streaming.py);
/// named here because this module is where the remedy for it is decided.
pub const GUARD_ABORT_FINISH_REASON: &str = "aborted_reasoning_overrun";

/// True when the guard cut this attempt, rather than the model choosing to stop.
///
/// This is the discriminator between the two model shapes that share a first
/// attempt. A model that WANTS more room stops on its own once it has enough and
/// the guard never fires; a model whose reasoning expands to fill the budget is cut
/// by the guard at every budget it is offered. So a guard abort on an ESCALATED
/// attempt means escalation cannot work for this model -- it was handed more and
/// spent more, without ever answering.
///
/// Deliberately NOT true of a plain `finish_reason=length` overrun: that one really
/// can be a budget shortage, and is the case the escalation exists to serve.
#[must_use]
pub fn reasoning_overrun_was_guard_aborted(finish_reason: &str) -> bool {
    finish_reason == GUARD_ABORT_FINISH_REASON
}

#[must_use]
pub fn reasoning_retry_budget(base_budget: u32) -> u32 {
    whole_u32(f64::from(base_budget) * REASONING_RETRY_MULTIPLIER).min(REASONING_RETRY_MAX_TOKENS)
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Diagnosis {
    pub category: &'static str,
    pub reason: String,
    pub evidence: String,
}

impl Diagnosis {
    const fn none() -> Self {
        Self {
            category: FAIL_NONE,
            reason: String::new(),
            evidence: String::new(),
        }
    }
}

/// HTTP 5xx | `server_overloaded` | inference capacity | service unavailable,
/// case-insensitive like the Python regex.
fn is_server_error(error: &str) -> bool {
    let lower = error.to_lowercase();
    lower.contains("http 5")
        || lower.contains("server_overloaded")
        || lower.contains("inference capacity")
        || lower.contains("service unavailable")
}

fn transport_failure(error: &str) -> Option<Diagnosis> {
    if error.contains("Model not found") {
        return Some(Diagnosis {
            category: FAIL_INFRA,
            reason: error.to_string(),
            evidence: "Model not loaded or wrong identifier".to_string(),
        });
    }
    if error.contains("Connection") {
        return Some(Diagnosis {
            category: FAIL_INFRA,
            reason: error.to_string(),
            evidence: "Server unreachable".to_string(),
        });
    }
    if is_server_error(error) {
        return Some(Diagnosis {
            category: FAIL_INFRA,
            reason: error.to_string(),
            evidence: "Server returned 5xx; the model never got to answer".to_string(),
        });
    }
    if error.contains("Timeout") || error.to_lowercase().contains("timed out") {
        return Some(Diagnosis {
            category: FAIL_TIMEOUT,
            reason: error.to_string(),
            evidence: "Model did not respond within the task timeout".to_string(),
        });
    }
    None
}

/// Classify why this attempt failed. Branch order mirrors the Python original
/// exactly, because the branches are ordered by "which mislabel hides the
/// truth best".
#[must_use]
pub fn classify_failure(
    error: Option<&str>,
    content: &str,
    reasoning: &str,
    finish_reason: &str,
    parsed: Option<&Value>,
    score: u8,
    parse_json: bool,
) -> Diagnosis {
    let _ = finish_reason;
    if score >= 90 {
        return Diagnosis::none();
    }

    // An empty error string matches nothing, exactly like Python's "" checks.
    let error = error.unwrap_or("");
    if let Some(d) = transport_failure(error) {
        return d;
    }

    // Before the parse branch: a reasoning model that never stopped returns
    // empty content, and on a JSON task that reads as "no JSON brackets".
    if content.is_empty() && !reasoning.is_empty() {
        return Diagnosis {
            category: FAIL_REASONING,
            reason: "Reasoned past the token budget".to_string(),
            evidence: format!(
                "{} chars of reasoning_content, empty content, finish_reason={}. \
                 Not a prompt-following failure: the model never stopped thinking, \
                 so the remedy is a budget change and not a prompt rewrite. WHICH \
                 change depends on the shape: a model whose reasoning scales with \
                 the TASK answers once the budget covers it, so raise max_tokens; a \
                 model whose reasoning expands to fill the BUDGET is guard-aborted \
                 at every budget and raising it only buys a longer failure. The run \
                 tells them apart by escalating once and reading the outcome.",
                reasoning.len(),
                finish_reason_or_unknown(finish_reason),
            ),
        };
    }

    if parse_json {
        return classify_json_failure(content, parsed, score);
    }

    if content.is_empty() {
        return Diagnosis {
            category: FAIL_FORMAT,
            reason: "Empty content".to_string(),
            evidence: "Model returned empty response".to_string(),
        };
    }

    let reasoning_markers = [
        "Let me",
        "I'll",
        "Wait,",
        "Actually,",
        "Here's my",
        "Thinking",
    ];
    let head: String = content.chars().take(200).collect();
    if reasoning_markers.iter().any(|m| head.contains(m)) && content.chars().count() > 200 {
        return Diagnosis {
            category: FAIL_FORMAT,
            reason: "Output starts with reasoning instead of a direct answer".to_string(),
            evidence: format!(
                "Model output {} chars starting with reasoning instead of a direct answer",
                content.chars().count()
            ),
        };
    }

    Diagnosis {
        category: FAIL_CONTENT,
        reason: "Output failed validation".to_string(),
        evidence: format!(
            "Output was {} chars but failed validation",
            content.chars().count()
        ),
    }
}

const fn finish_reason_or_unknown(finish_reason: &str) -> &str {
    if finish_reason.is_empty() {
        "unknown"
    } else {
        finish_reason
    }
}

/// The `parse_json` branches of `failures.py::classify_failure`: no brackets
/// at all is a FORMAT failure, brackets the parser could not use is PARSE,
/// a long prose preamble before the first bracket is FORMAT when the score
/// collapsed (the reasoning likely ate the context) and CONTENT otherwise.
fn classify_json_failure(content: &str, parsed: Option<&Value>, score: u8) -> Diagnosis {
    let raw_len = content.len();
    let first_bracket = ['[', '{']
        .iter()
        .filter_map(|c| content.find(*c))
        .min()
        .unwrap_or(raw_len);
    let has_json_chars = first_bracket < raw_len;
    let has_prose_before_json = has_json_chars && first_bracket > 200;
    match (parsed.is_some(), has_json_chars) {
        (false, false) => Diagnosis {
            category: FAIL_FORMAT,
            reason: "No JSON in output".to_string(),
            evidence: format!("Output was {raw_len} chars of prose with no JSON brackets"),
        },
        (false, true) => Diagnosis {
            category: FAIL_PARSE,
            reason: "JSON extraction failed".to_string(),
            evidence: format!(
                "Output had JSON-like content at char {first_bracket} of {raw_len} but parser \
                 couldn't extract valid JSON"
            ),
        },
        (true, _) if has_prose_before_json => {
            let evidence = format!(
                "Model emitted {first_bracket} chars of reasoning before first JSON bracket \
                 (total {raw_len} chars)"
            );
            if score < 50 {
                Diagnosis {
                    category: FAIL_FORMAT,
                    reason: "parse_json task scored below threshold".to_string(),
                    evidence: format!("{evidence}; reasoning may have consumed the context window"),
                }
            } else {
                Diagnosis {
                    category: FAIL_CONTENT,
                    reason: "parse_json task scored below threshold".to_string(),
                    evidence,
                }
            }
        }
        (true, _) => Diagnosis {
            category: FAIL_CONTENT,
            reason: "parse_json task scored below threshold".to_string(),
            evidence: String::new(),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn diag(error: Option<&str>, content: &str, reasoning: &str, score: u8) -> Diagnosis {
        classify_failure(error, content, reasoning, "stop", None, score, false)
    }

    #[test]
    fn infra_timeout_and_reasoning_are_distinguished() {
        assert_eq!(
            diag(Some("Connection failed - is server running?"), "", "", 0).category,
            FAIL_INFRA
        );
        assert_eq!(
            diag(Some("HTTP 503: Server is at inference capacity"), "", "", 0).category,
            FAIL_INFRA
        );
        assert_eq!(diag(Some("Timeout"), "", "", 0).category, FAIL_TIMEOUT);
        assert_eq!(
            diag(None, "", "thinking ".repeat(300).trim(), 0).category,
            FAIL_REASONING
        );
    }

    #[test]
    fn reasoning_beats_format_on_empty_content_with_thinking() {
        // The mislabel one level down: empty content + reasoning must be
        // REASONING even though content is also (trivially) "empty".
        let d = diag(None, "", "a very long chain of thought indeed", 0);
        assert_eq!(d.category, FAIL_REASONING);
        assert!(d.evidence.contains("never stopped thinking"));
    }

    #[test]
    fn empty_and_prosey_outputs_are_format_not_content() {
        assert_eq!(diag(None, "", "", 0).category, FAIL_FORMAT);
        let prosey = format!("Let me think about this. {}", "word ".repeat(60));
        assert_eq!(diag(None, &prosey, "", 0).category, FAIL_FORMAT);
    }

    #[test]
    fn real_output_that_fails_validation_is_content() {
        assert_eq!(
            diag(None, "a short but wrong answer", "", 0).category,
            FAIL_CONTENT
        );
    }

    #[test]
    fn a_passing_score_is_never_a_failure_even_with_an_error_string() {
        assert_eq!(diag(Some("Timeout"), "answer", "", 95).category, FAIL_NONE);
    }

    #[test]
    fn retry_budget_grows_but_is_bounded() {
        assert_eq!(reasoning_retry_budget(32_000), 64_000);
        assert_eq!(reasoning_retry_budget(10_000), 20_000);
        assert_eq!(reasoning_retry_budget(40_000), 64_000);
    }

    #[test]
    fn json_task_failures_are_classified_like_python() {
        let d = classify_failure(None, "just prose, no brackets", "", "stop", None, 10, true);
        assert_eq!(d.category, FAIL_FORMAT);
        assert_eq!(d.reason, "No JSON in output");

        let d = classify_failure(None, "here: {broken json", "", "stop", None, 10, true);
        assert_eq!(d.category, FAIL_PARSE);
        assert!(d.evidence.contains("at char 6 of 18"), "{}", d.evidence);

        let preamble = format!("{}{{\"a\": 1}}", "reasoning ".repeat(30));
        let parsed = serde_json::json!({"a": 1});
        let d = classify_failure(None, &preamble, "", "stop", Some(&parsed), 20, true);
        assert_eq!(d.category, FAIL_FORMAT);
        assert!(d.evidence.contains("consumed the context window"));
        let d = classify_failure(None, &preamble, "", "stop", Some(&parsed), 70, true);
        assert_eq!(d.category, FAIL_CONTENT);
        assert!(d
            .evidence
            .contains("chars of reasoning before first JSON bracket"));

        let d = classify_failure(None, "{\"a\": 1}", "", "stop", Some(&parsed), 40, true);
        assert_eq!(d.category, FAIL_CONTENT);
        assert!(d.evidence.is_empty());
    }
}
